### ISSUE TITLE
Drug Response VAE

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#620]
https://arxiv.org/abs/1706.08203

Looks interesting. We should add an implementation to DeepChem

**Comment by LRParser** [comment]
Agreed, it doesn't look like any of their code is open source though. Would a Keras implementation be the desired way to go? Ref: https://github.com/fchollet/keras/blob/master/examples/variational_autoencoder.py

**Comment by rbharath** [comment]
We dropped Keras support a while ago due the rapidly changing API which broke the build a few times. Has the Keras API stabilized now?

It might also be straightforward to do this in TensorGraph (cc @lilleswing)

**Comment by lilleswing** [comment]
The main network topology is described in this paper -- https://papers.nips.cc/paper/6581-improved-variational-inference-with-inverse-autoregressive-flow.pdf

![selection_039](https://user-images.githubusercontent.com/1225605/27614111-90947978-5b6d-11e7-9e91-0ed223b09061.png)

I had trouble following their semi-supervised learning scheme so I don't know if I can comment on how easy it would be to port.

@LRParser the code you linked is already ported into deepchem here
Gene Expressions are expressed in a bitvector really easily -- but small molecules are not.  This is the main challenge for creating a molecular autoencoder.

https://github.com/deepchem/deepchem/tree/master/deepchem/models/autoencoder_models

**Comment by lilleswing** [comment]
Their semi-supervised learning scheme works because they can train both x_1 and x_2 over the encoder without training the gene expression changes as shown here.

![selection_040](https://user-images.githubusercontent.com/1225605/27614450-d4bf7fa6-5b6f-11e7-8cec-c2becdf0cedf.png)

This should be doable to port to Tensorgraph.  Getting our hands on the datasets to prove that it worked might be the trickiest part.

**Comment by rbharath** [comment]
I'd love to get some gene expression data in DeepChem. There are some very impressive results here, notably:

http://stm.sciencemag.org/content/3/96/96ra77

If we could get that implemented in DeepChem, would be amazing!

**Comment by LRParser** [comment]
Ok, this one is over my head for the moment, but I'll do my best to follow along and learn Tensorgraph.

**Comment by rbharath** [comment]
This would be a good intermediate level contribution, especially for those with prior deep learning experience.
